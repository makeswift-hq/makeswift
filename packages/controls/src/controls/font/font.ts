import { z } from 'zod'

import { safeParse, type ParseResult } from '../../lib/zod'

import {
  type DeserializedRecord,
  type SerializedRecord,
} from '../../serialization'

import { ControlDefinition, serialize, type Resolvable } from '../definition'
import { DefaultControlInstance, type SendMessage } from '../instance'

type Schema = typeof Definition.schema
type DataType = z.infer<Schema['data']>
type ValueType = z.infer<Schema['value']>
type ResolvedValueType = z.infer<Schema['resolvedValue']>
type InstanceType = DefaultControlInstance

class Definition extends ControlDefinition<
  typeof Definition.type,
  unknown,
  DataType,
  ValueType,
  ResolvedValueType,
  InstanceType
> {
  static readonly type = 'makeswift::controls::font' as const

  static get schema() {
    const type = z.literal(Definition.type)

    const data = z.object({
      fontFamily: z.string(),
      fontStyle: z.string(),
      fontWeight: z.number(),
    })

    const value = data
    const resolvedValue = data

    const definition = z.object({
      type,
    })

    return {
      type,
      definition,
      data,
      value,
      resolvedValue,
    }
  }

  static deserialize(data: DeserializedRecord): Definition {
    if (data.type !== Definition.type) {
      throw new Error(
        `Font: expected type ${Definition.type}, got ${data.type}`,
      )
    }

    Definition.schema.definition.parse(data)
    return new (class Font extends Definition {})()
  }

  constructor() {
    super({})
  }

  get controlType() {
    return Definition.type
  }

  get schema() {
    return Definition.schema
  }

  safeParse(data: unknown | undefined): ParseResult<DataType | undefined> {
    return safeParse(this.schema.data, data)
  }

  fromData(data: DataType | undefined): ValueType | undefined {
    return this.schema.data.optional().parse(data)
  }

  toData(value: ValueType): DataType {
    return value
  }

  copyData(data: DataType | undefined): DataType | undefined {
    if (data == null) return data

    return {
      ...data,
    }
  }

  resolveValue(
    data: DataType | undefined,
  ): Resolvable<ResolvedValueType | undefined> {
    return {
      readStableValue: () => this.fromData(data),
      subscribe: () => () => {},
    }
  }

  createInstance(sendMessage: SendMessage) {
    return new DefaultControlInstance(sendMessage)
  }

  serialize(): [SerializedRecord, Transferable[]] {
    return serialize(this.config, {
      type: Definition.type,
    })
  }
}

export class FontDefinition extends Definition {}

export function Font(): FontDefinition {
  return new FontDefinition()
}
