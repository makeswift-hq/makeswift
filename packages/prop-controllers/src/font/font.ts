import { P, match } from 'ts-pattern'
import {
  ControlDataTypeKey,
  Options,
  Types,
  createResponsiveValueSchema,
} from '../prop-controllers'
import { z } from 'zod'

const fontDataSchema = z.string()

export type FontData = z.infer<typeof fontDataSchema>

const responsiveFontDataSchema = createResponsiveValueSchema(fontDataSchema)

export type ResponsiveFontData = z.infer<typeof responsiveFontDataSchema>

const fontPropControllerDataV0Schema = responsiveFontDataSchema

export type FontPropControllerDataV0 = z.infer<
  typeof fontPropControllerDataV0Schema
>

export const FontPropControllerDataV1Type = 'prop-controllers::font::v1'

const fontPropControllerDataV1Schema = z.object({
  [ControlDataTypeKey]: z.literal(FontPropControllerDataV1Type),
  value: responsiveFontDataSchema,
})

export type FontPropControllerDataV1 = z.infer<
  typeof fontPropControllerDataV1Schema
>

export const fontPropControllerDataSchema = z.union([
  fontPropControllerDataV0Schema,
  fontPropControllerDataV1Schema,
])

export type FontPropControllerData = z.infer<
  typeof fontPropControllerDataSchema
>

type FontOptions = Options<{ preset?: FontPropControllerData; label?: string }>

type FontDescriptorV0<
  _T = FontPropControllerDataV0,
  U extends FontOptions = FontOptions,
> = {
  type: typeof Types.Font
  options: U
}

type FontDescriptorV1<
  _T = FontPropControllerData,
  U extends FontOptions = FontOptions,
> = {
  type: typeof Types.Font
  version: 1
  options: U
}

export type FontDescriptor<
  _T = FontPropControllerData,
  U extends FontOptions = FontOptions,
> = FontDescriptorV0<_T, U> | FontDescriptorV1<_T, U>

export type ResolveFontPropControllerValue<T extends FontDescriptor> =
  T extends FontDescriptor ? ResponsiveFontData | undefined : never

/**
 * @deprecated Imports from `@makeswift/prop-controllers` are deprecated. Use
 * `@makeswift/runtime/controls` instead.
 */
export function Font(
  options: FontOptions = {},
): FontDescriptor<FontPropControllerData> {
  return { type: Types.Font, version: 1, options }
}

export function getFontPropControllerDataResponsiveFontData(
  data: FontPropControllerData | undefined,
): ResponsiveFontData | undefined {
  return match(data)
    .with(
      { [ControlDataTypeKey]: FontPropControllerDataV1Type },
      (v1) => v1.value,
    )
    .otherwise((v0) => v0)
}

export function createFontPropControllerDataFromResponsiveFontData(
  responsiveFontData: ResponsiveFontData,
  definition?: FontDescriptor,
): FontPropControllerData {
  return match(definition)
    .with(
      { version: 1 },
      P.nullish,
      () =>
        ({
          [ControlDataTypeKey]: FontPropControllerDataV1Type,
          value: responsiveFontData,
        } as const),
    )
    .otherwise(() => responsiveFontData)
}
