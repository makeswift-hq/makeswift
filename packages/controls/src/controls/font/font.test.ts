import { Font, FontDefinition } from './font'

describe('Font', () => {
  describe('constructor', () => {
    test('returns correct definition', () => {
      expect(Font()).toMatchSnapshot()
    })

    test('disallows extraneous properties', () => {
      // @ts-expect-error
      Font({
        extra: 'extra',
      })
    })
  })

  describe('assignability', () => {
    function assignTest(_def: FontDefinition) {}
    assignTest(Font())
  })
})
