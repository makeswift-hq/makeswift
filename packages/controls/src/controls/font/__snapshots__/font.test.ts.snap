// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Font constructor returns correct definition 1`] = `
FontDefinition {
  "config": {},
}
`;
